atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidatorTest.java

```diff
@@ -0,0 +1,294 @@
+        assertThat(validator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        allHosts.keySet(),
+                        allHosts,
+                        HumanReadableDuration.milliseconds(1),
+                        HumanReadableDuration.milliseconds(1)))
+                .isNotEmpty();
+        assertThat(metrics.validationFailures().getCount()).isEqualTo(1);
+        assertThat(metrics.validationLatency().getCount()).isEqualTo(1);
+    }
+
+    @Test
+    public void returnsEmptyWhenGivenNoServers() {
+        assertThat(validator.getNewHostsWithInconsistentTopologies(ImmutableSet.of(), ImmutableMap.of()))
+                .isEmpty();
+    }
+
+    @Test
+    public void throwsWhenAllHostsDoesNotContainNewHosts() {
+        assertThatThrownBy(() -> validator.getNewHostsWithInconsistentTopologies(
+                        ImmutableSet.of(createCassandraServer("foo")), ImmutableMap.of()))
+                .isInstanceOf(SafeIllegalArgumentException.class);
+    }
+
+    @Test
+    public void returnsEmptyWhenOnlyNewServersAndAllMatch() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(newHosts);
+        setHostIds(allHosts.values(), DEFAULT_RESULT);
+        assertThat(validator.getNewHostsWithInconsistentTopologies(allHosts.keySet(), allHosts))
+                .isEmpty();
+    }
+
+    @Test
+    public void returnsNewHostsWhenOnlyNewServersAndOneMismatch() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(newHosts);
+        setHostIds(
+                filterContainers(allHosts, server -> !server.equals(newHostOne)),
+                HostIdResult.success(ImmutableSet.of("uuid1", "uuid2")));
+        setHostIds(
+                filterContainers(allHosts, newHostOne::equalsIgnoreCase),
+                HostIdResult.success(ImmutableSet.of("uuid3", "uuid2")));
+        assertThat(validator.getNewHostsWithInconsistentTopologies(allHosts.keySet(), allHosts))
+                .containsExactlyElementsOf(allHosts.keySet());
+    }
+
+    @Test
+    public void returnsNewHostsWhenNewServersMismatchOldServers() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(Sets.union(newHosts, oldHosts));
+        Set<CassandraServer> newServers = filterServers(allHosts, newHosts::contains);
+        setHostIds(
+                filterContainers(allHosts, newHosts::contains),
+                HostIdResult.success(ImmutableSet.of("uuid1", "uuid2")));
+        setHostIds(
+                filterContainers(allHosts, oldHosts::contains),
+                HostIdResult.success(ImmutableSet.of("uuid3", "uuid2")));
+        assertThat(validator.getNewHostsWithInconsistentTopologies(newServers, allHosts))
+                .containsExactlyElementsOf(newServers);
+    }
+
+    @Test
+    public void validateNewlyAddedHostsReturnsOnlyMismatchingNewHosts() {
```

Comment:
What does this test that the above 2 don't?

Reply:
yeah naming here is hard. 
`returnsNewHostsWhenOnlyNewServersAndOneMismatch` -> If this is my first validation (hence only new servers), and one mismatches, we should fail. 

`returnsNewHostsWhenNewServersMismatchOldServers` -> if new servers have consensus on host ids [1,2], but old servers have consensus on [2,3], then return all new servers as this mismatch new servers. 

`validateNewlyAddedHostsReturnsOnlyMismatchingNewHosts` -> if one new server has host ids [1,2], but the other two have the same host ids as the current servers [2, 3], we only expect the the one new server to be returned